src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.CacheEntry.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers.ImportCompletion
+{
+    internal abstract partial class AbstractTypeImportCompletionService
+    {
+        private readonly struct CacheEntry
+        {
+            public string Language { get; }
+
+            public Checksum Checksum { get; }
+
+            private ImmutableArray<TypeImportCompletionItemInfo> ItemInfos { get; }
+
+            private CacheEntry(
+                Checksum checksum,
+                string language,
+                ImmutableArray<TypeImportCompletionItemInfo> items)
+            {
+                Checksum = checksum;
+                Language = language;
+
+                ItemInfos = items;
+            }
+
+            public ImmutableArray<CompletionItem> GetItemsForContext(
+                string language,
+                string genericTypeSuffix,
+                bool isInternalsVisible,
+                bool isAttributeContext,
+                bool isCaseSensitive)
+            {
+                var isSameLanguage = Language == language;
+                if (isSameLanguage && !isAttributeContext)
```

Comment:
doc all this logic. (including the loops below).

Reply:
Done

Reply:
did you not push?